R/drive_id-class.R

```diff
@@ -0,0 +1,162 @@
+new_drive_id <- function(x = character()) {
+  vec_assert(x, character())
+  new_vctr(x, class = "drive_id")
+}
+
+validate_drive_id <- function(x) {
+  ok <- is_valid_drive_id(x)
+  if (all(ok)) {
+    return(x)
+  }
+
+  # proceed with plain character vector
+  x <- unclass(x)
+  # pragmatism re: how to cli-style a path that is the empty string
+  x[!nzchar(x)] <- "\"\""
+
+  drive_abort(c(
+    "A {.cls drive_id} must match this regular expression: \\
+     {.code {drive_id_regex()}}",
+    "Invalid input{?s}:{cli::qty(sum(!ok))}",
+    bulletize(gargle_map_cli(x[!ok]), bullet = "x")
+  ))
+}
+
+drive_id_regex <- function() "^[a-zA-Z0-9_-]+$"
+
+is_valid_drive_id <- function(x) {
+  # among practitioners, It Is Known that file IDs have >= 25 characters
+  # but I'm not convinced the pros outweigh the cons re: checking length
+  # for example, in tests, it's nice to not worry about this
+  grepl(drive_id_regex(), x) | is.na(x)
+}
+
+is_drive_id <- function(x) {
+  inherits(x, "drive_id")
+}
+
+#' @export
+gargle_map_cli.drive_id <- function(x, ...) {
+  # x <- unclass(x)
+  # NextMethod()
+  # TODO: I know this is gross, but I cannot figure out any "nice" way
```

Comment:
Re-review after adding `character` to class; use `sloop::s3_dispatch()` to diagnose.

Reply:
Using `inherit_base_class = TRUE` (and thereby adding `character` to the class vector "fixed" this. Non-gross `NextMethod()` works now.